### PR TITLE
feat: smart contract with improved dx

### PR DESCRIPTION
**Pull request by kratico**
## Description

Add a smart contract deployment and call example.

The example showcase how to
- send non-mutable messages
- send mutable messages
- encode message arguments and decode message return values using the contract metadata
- decode message events

### CODE REVIEW

File: examples/smart_contract.ts [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+      "ContractsApi_call",
+      callData,
+    )
+      .next((encodedResponse) => {
+        return $contractsApiCallResult.decode(U.hex.decode(encodedResponse))
+      })
+  }
+}
+
+const flipperContract = new SmartContract(contract.metadata, contractAddress)
```
Comment (kratico) [inline examples/smart_contract.ts:246]: The following contract calls could be simplified without `console.logs` like
```
await flipperContract.call(T.alice.publicKey, "get", []).run()
await flipperContract.tx(T.alice.publicKey, "flip", [], T.alice.sign).run()
await flipperContract.call(T.alice.publicKey, "get", []).run()
await flipperContract.call(T.alice.publicKey, "get_count", []).run()
await flipperContract.tx(T.alice.publicKey, "inc", [], T.alice.sign).run()
await flipperContract.tx(T.alice.publicKey, "inc", [], T.alice.sign).run()
await flipperContract.call(T.alice.publicKey, "get_count", []).run()
await flipperContract.tx(T.alice.publicKey, "inc_by", [3], T.alice.sign).run()
await flipperContract.call(T.alice.publicKey, "get_count", []).run()
await flipperContract.call(T.alice.publicKey, "m1", [2, true]).run()
await flipperContract.call(T.alice.publicKey, "m1", [3, false]).run()
```

For messages that mutate the state, `contract.tx(origin, message, args, sign)` is used.
For other messages, `contract.call(origin, message, args)` is used.

The source code for this contract is
```
#[ink::contract]
mod flipper {

    /// Defines the storage of your contract.
    /// Add new fields to the below struct in order
    /// to add new static storage fields to your contract.
    #[ink(storage)]
    pub struct Flipper {
        /// Stores a single `bool` value on the storage.
        value: bool,
        /// Stores count
        count: i32,
    }

    #[derive(Debug, scale::Encode, scale::Decode)]
    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
    pub struct M2 {
        a: u32,
        b: bool,
    }

    impl Flipper {
        /// Constructor that initializes the `bool` value to the given `init_value`.
        #[ink(constructor)]
        pub fn new(init_value: bool) -> Self {
            Self {
                value: init_value,
                count: 0,
            }
        }

        /// Constructor that initializes the `bool` value to `false`.
        ///
        /// Constructors can delegate to other constructors.
        #[ink(constructor)]
        pub fn default() -> Self {
            Self::new(Default::default())
        }

        /// A message that can be called on instantiated contracts.
        /// This one flips the value of the stored `bool` from `true`
        /// to `false` and vice versa.
        #[ink(message)]
        pub fn flip(&mut self) {
            self.value = !self.value;
        }

        /// Simply returns the current value of our `bool`.
        #[ink(message)]
        pub fn get(&self) -> bool {
            self.value
        }

        /// multiple arg method1.
        #[ink(message)]
        pub fn m1(&self, a: u32, b: bool) -> (u32, bool) {
            (a, b)
        }

        /// multiple arg method2.
        #[ink(message)]
        pub fn m2(&self, a: u32, b: bool) -> M2 {
            M2 { a, b }
        }

        /// get the current count.
        #[ink(message)]
        pub fn get_count(&self) -> i32 {
            self.count
        }

        /// increment current count by 1
        #[ink(message)]
        pub fn inc(&mut self) {
            self.inc_by(1);
        }

        /// decrement current count by 1
        #[ink(message)]
        pub fn dec(&mut self) {
            self.inc_by(-1);
        }

        /// increment current count by n
        #[ink(message)]
        pub fn inc_by(&mut self, n: i32) {
            self.count += n;
        }
    }
```

File: examples/smart_contract.ts [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+const contractAddress = U.throwIfError(await instantiateContractTx().run())
+console.log("Deployed Contract address", U.ss58.encode(42, contractAddress))
+
+function instantiateContractTx() {
+  const constructor = findContractConstructorByLabel("default")!
+  const salt = Uint8Array.from(Array.from([0, 0, 0, 0]), () => Math.floor(Math.random() * 16))
+  const value = preSubmitContractInstantiateDryRunGasEstimate(constructor, contract.wasm, salt)
+    .next(({ gasRequired, result: { accountId } }) => {
+      // the contract address derived from the code hash and the salt
+      console.log("Derived contract address", U.ss58.encode(42, accountId))
```
Comment (harrysolovay) [inline examples/smart_contract.ts:30]: We can grab the prefix from the chain metadata (a constant)

File: examples/smart_contract.ts [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+      return {
+        type: "instantiateWithCode",
+        value: 0n,
+        gasLimit: {
```
Comment (harrysolovay) [inline examples/smart_contract.ts:34]: Why not just `gasLimit: gasRequired`?

File: examples/smart_contract.ts [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+      }
+      return
+    }
+  )
+  return C.events(tx, finalizedIn).next((events) => {
+    const extrinsicFailed = events.some((e) =>
+      e.event?.type === "System" && e.event?.value?.type === "ExtrinsicFailed"
+    )
+    if (extrinsicFailed) {
+      return new Error("extrinsic failed")
```
Comment (harrysolovay) [inline examples/smart_contract.ts:67]: How might we determine the cause?

Comment (kratico) [reply examples/smart_contract.ts]: created an `ExtrinsicFailed` error and added the extrinsic failed event as the error `cause`

**Review by harrysolovay — APPROVED** [review]

**Review by harrysolovay — APPROVED** [review]

**Pull request merged**